File: C/Binary_Search.c
#include<stdio.h>
int  main()
{
	int a[10],i,n,m,c=0,l,u,mid;
	printf("enter the size of an array");
	scanf("%d",&n);
	printf("enter the element in ascending order");
	for(i=0;i<n;i++)
    {
		scanf("%d",&a[i]);
   	}
	printf("enter the number to be searched");
	scanf("%d",&m);
	l=0;
	u=n-1;
	while(l<=u)
	{
		mid=(l+u)/2;
		if(m==a[mid])
		{
			c=1;
			break;
		}
		else if(m<a[mid])
		{
			u=mid-1;
		}
		else
		{
			l=mid+1;
		}
	}
	if(c==0)
	{
		printf("the number is not found");
	}
	else
	{
		printf("the number is found");
		return 0;
	}
}
	

File: Rust/Binary Search.rs
//intermediate
//The function call would be --> binary_search(&"element to search",&vec![..,array to search the element in,..]);
//index of the searched element would be returned

pub fn binary_search<T: PartialEq + PartialOrd>(item: &T, array: &[T]) -> i32 {
    let mut pos = -1; // value for not found

    if array.is_empty() {
        return pos;
    }

    let mut left = 0;
    let mut right = array.len() - 1;

    while left < right {
        let mid = left + (right - left) / 2;

        if &array[mid] > item {
            right = mid - 1;
        } else if &array[mid] < item {
            left = mid + 1;
        } else {
            left = mid;
            break;
        }
    }

    if &array[left] == item {
        	pos = left as i32;
        return pos;
    } else {
        return pos;
    }
}
